### DISCUSSION TITLE
how to start with rdkit in c++

### DISCUSSION THREAD

**Discussion by rmrmg** [rdkit/rdkit#3734] [Q&A] [answered]
Here is https://rdkit.org/docs/GettingStartedInC++.html tutorial how to start with RDKit in C++ but there is nothing about how to set environment/compilation flags, so I started with conda env, g++ and example1.cpp from https://github.com/rdkit/rdkit/blob/master/Docs/Book/C%2B%2BExamples/example1.cpp 
My first trial with c++14, specified include dir to rdkit and boost as well as  lib dir
` g++ -std=c++14 -I/home/rmrmg/miniconda3/envs/rdkit2020Q1/include -I/home/rmrmg/miniconda3/envs/rdkit2020Q1/include/rdkit   -L/home/rmrmg/miniconda3/envs/rdkit2020Q1/lib/  example1.cpp  -o test`
gives following error:

>/usr/bin/ld: /tmp/ccWd8ZU4.o: in function `main':
example1.cpp:(.text+0xaa): undefined reference to `RDKit::ROMol::getNumAtoms(bool) const'
/usr/bin/ld: example1.cpp:(.text+0x116): undefined reference to `RDKit::ROMol::getNumAtoms(bool) const'
/usr/bin/ld: example1.cpp:(.text+0x1cf): undefined reference to `RDKit::MolFileToMol(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, bool, bool, bool)'
/usr/bin/ld: /tmp/ccWd8ZU4.o: in function `RDKit::SmilesToMol(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, int, bool, std::map<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::less<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, std::allocator<std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > > >*)':
example1.cpp:(.text._ZN5RDKit11SmilesToMolERKNSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEEibPSt3mapIS5_S5_St4lessIS5_ESaISt4pairIS6_S5_EEE[_ZN5RDKit11SmilesToMolERKNSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEEibPSt3mapIS5_S5_St4lessIS5_ESaISt4pairIS6_S5_EEE]+0x70): undefined reference to `RDKit::SmilesToMol(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, RDKit::SmilesParserParams const&)'
collect2: error: ld returned 1 exit status

It seems that ld cannot fine RDKit library to link with so I tried several option and finally found one which works:

`g++ -std=c++14 -I/home/rmrmg/miniconda3/envs/rdkit2020Q1/include -I/home/rmrmg/miniconda3/envs/rdkit2020Q1/include/rdkit   -L/home/rmrmg/miniconda3/envs/rdkit2020Q1/lib -lRDKitSmilesParse -lRDKitGraphMol -lRDKitDataStructs -lRDKitFileParsers example1.cpp  -o test`

And it works!

When I started writing I was sure this will end with question but fortunately I did it. If you want comment how can I do this better (trying all lib and checking what happen is probably slightly suboptimal) please do.

**Comment by greglandrum** [comment, marked as answer]
We don't currently have an easy way to know which library you should be linking against aside from the general rule that if you are #include'ing something  you probably need to link against the corresponding library.

FWIW: Your final example command line matches more or less what I would do.

**Comment by greglandrum** [comment]
I believe that this has been answered, so I'm marking it as such.
If you disagree with that, please remove the "answered" marker or followup here.
